plugins/vli/fu-vli-pd-parade-device.c

```diff
@@ -0,0 +1,668 @@
+					    buf, bufsz, NULL,
+					    FU_VLI_DEVICE_TIMEOUT,
+					    NULL, error)) {
+		g_prefix_error (error, "failed to read 0x%x:0x%x: ", page_addr, reg_offset);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_vli_pd_parade_device_i2c_write (FuVliPdParadeDevice *self,
+				   guint8 page_addr,
+				   guint8 reg_offset,	/* customers addr offset */
+				   guint8 val,		/* only one byte supported */
+				   GError **error)
+{
+	guint16 value;
+	guint16 index;
+	guint8 buf[2] = { 0x0 };	/* apparently unused... */
+
+	/* page2_address = 0x14/0x94, page7_address = 0x1E/0x9E --
+	 * but VL103 FW only Use bits[7:1], so divide by 2 */
+	value = ((guint16) reg_offset << 8) | (page_addr >> 1);
+	index = (guint16) val << 8;
+	if (!g_usb_device_control_transfer (fu_usb_device_get_dev (FU_USB_DEVICE (self)),
+					    G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					    G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					    G_USB_DEVICE_RECIPIENT_DEVICE,
+					    0xA6, value, index,
+					    buf, 0x0, NULL,
+					    FU_VLI_DEVICE_TIMEOUT,
+					    NULL, error)) {
+		g_prefix_error (error, "failed to write 0x%x:0x%x: ", page_addr, reg_offset);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_vli_pd_parade_device_stop_mcu (FuVliPdParadeDevice *self, GError **error)
+{
+	if (!fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0xBC, 0xC0, error))
+		return FALSE;
+	if (!fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0xBC, 0x40, error))
+		return FALSE;
+	return TRUE;
+}
+
+static gboolean
+fu_vli_pd_parade_device_set_offset (FuVliPdParadeDevice *self, guint16 addr, GError **error)
+{
+	if (!fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0x8E, addr >> 8, error))
+		return FALSE;
+	if (!fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0x8F, addr & 0xff, error))
+		return FALSE;
+	return TRUE;
+}
+
+static gboolean
+fu_vli_pd_parade_device_read_fw_ver (FuVliPdParadeDevice *self, GError **error)
+{
+	guint8 buf[0x20] = { 0x0 };
+	g_autofree gchar *version_str = NULL;
+
+	/* Stop_MCU */
+	if (!fu_vli_pd_parade_device_stop_mcu (self, error))
+		return FALSE;
+	if (!fu_vli_pd_parade_device_set_offset (self, 0x0, error))
+		return FALSE;
+	g_usleep (1000 * 10);
+	if (!fu_vli_pd_parade_device_i2c_read (self, self->page7_addr, 0x02, buf, 0x1, error))
+		return FALSE;
+	if ((buf[0] != 0x01) && (buf[0] != 0x02)) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "not supported");
+		return FALSE;
+	}
+
+	g_debug ("getting FW%X version", buf[0]);
+	if (!fu_vli_pd_parade_device_set_offset (self, 0x5000 | buf[0], error))
+		return FALSE;
+	if (!fu_vli_pd_parade_device_i2c_read (self, self->page7_addr, 0x00, buf, sizeof(buf), error))
+		return FALSE;
+
+	/* format as a hex triplet */
+	version_str = g_strdup_printf ("%02X.%02X.%02X", buf[0], buf[1], buf[2]);
+	fu_device_set_version (FU_DEVICE (self), version_str, FWUPD_VERSION_FORMAT_TRIPLET);
+	return TRUE;
```

Comment:
since you stopped MCU at start, you don't need to restart the MCU after reading FW version?

Reply:
Valid question: waiting for VIA about this.

Reply:
We now have a `fu_vli_pd_parade_device_start_mcu()` -- thanks.